### PR TITLE
Validate network parameters and ledger index

### LINKED ISSUE #35: Check the network ID the database was created for at startup
After chaning the networkID in the alphanet and resetting it, the milestone index in the indexer was bigger then the current milestone in the node.

That caused that ledger updates where never applied.

We should add a check if the database the indexer is using fits the network hornet is operating on.
For this we need to store the network id in the database.

### PR DESCRIPTION
**Pull request by oopsmonk**
This PR fixes #35 
The indexer checks network parameters and the ledger index to decide if drop the outdated database.
drop database if
- network name and protocol version has been changed
- network ledger index is smaller than the indexer index, which means the network has been reset.

### CODE REVIEW

File: pkg/indexer/types.go [hunk 1]
```diff
@@ -30,6 +30,16 @@ type status struct {
...
 }
 
+type protocol struct {
+	ID uint `gorm:"primaryKey;notnull"`
+	// The version of the protocol running
+	Version byte
+	// The network name
+	NetworkName string
+	// The HRP prefix used for Bech32 addresses in the network
+	Bech32HRP string
```
Comment (muXxer) [inline pkg/indexer/types.go]: We don't need the Bech32HRP I guess.

File: pkg/indexer/types.go [hunk 2]
```diff
@@ -30,6 +30,16 @@ type status struct {
 	LedgerIndex uint32
 }
 
+type protocol struct {
+	ID uint `gorm:"primaryKey;notnull"`
+	// The version of the protocol running
+	Version byte
```
Comment (muXxer) [inline pkg/indexer/types.go]: We don't need an entry for the version.
It should instead check if the Version we receive is supported by this plugin.

File: pkg/indexer/types.go [hunk 3]
```diff
@@ -30,6 +30,16 @@ type status struct {
...
+	// The network name
+	NetworkName string
```
Comment (muXxer) [inline pkg/indexer/types.go]: Instead of creating the protocol table, you can simply add this to the status table.

File: pkg/indexer/indexer.go [hunk 4]
```diff
@@ -342,6 +345,61 @@ func (i *Indexer) LedgerIndex() (uint32, error) {
...
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			// database is empty, create the protocol table
+			return false, i.CreateProtocolTable(networkProtocol)
+		}
+		return false, err
+	}
+
+	// should keep or drop current database?
+	// TODO: should check Bech32HRP too?
+	if networkProtocol.Version != currentProtocol.Version {
```
Comment (muXxer) [inline pkg/indexer/indexer.go]: Do not drop the database in that case, but shut down inx-indexer.

File: core/indexer/component.go [hunk 5]
```diff
@@ -173,14 +180,30 @@ func checkIndexerState(ctx context.Context) (uint32, error) {
...
 		CoreComponent.LogInfo("Re-import initial ledger...")
 
 		if err := deps.Indexer.Clear(); err != nil {
 			return 0, fmt.Errorf("clearing Indexer failed! Error: %w", err)
 		}
 		needsToFillIndexer = true
+		needsCreateProtocolTable = true
+	}
+
+	if needsCreateProtocolTable {
```
Comment (muXxer) [inline core/indexer/component.go]: Instead of creating the protocol table, we can add that to the state table.

Comment (oopsmonk) [reply core/indexer/component.go]: alright, I can add it to the state table since only the network id is needed.

File: pkg/indexer/types.go (hunk 3 above)
Comment (oopsmonk) [reply pkg/indexer/types.go]: For keeping the logic clear and extensible, we should not merge it with the status table.

File: core/indexer/component.go [hunk 6]
```diff
@@ -155,53 +160,72 @@ func newEcho() *echo.Echo {
...
+
+	protocolParams := deps.NodeBridge.ProtocolParameters()
+	// check protocol version
+	if protocolParams.Version != IndexerProtocolVersion {
+		return nil, fmt.Errorf("the supported protocol version is %d but the node protocol is %d", IndexerProtocolVersion, protocolParams.Version)
+	}
+
+	nodeStatus, err := deps.NodeBridge.NodeStatus()
+	if err != nil {
+		return nil, fmt.Errorf("error loading pruning index: %s", err)
```
Comment (muXxer) [inline core/indexer/component.go]:
```suggestion
		return nil, fmt.Errorf("error loading node status: %s", err)
```

**Review by muXxer — APPROVED** [review]

**Pull request merged**